src/libcore/iter/range.rs

```diff
@@ -5,169 +5,288 @@ use crate::usize;
 
 use super::{FusedIterator, TrustedLen};
 
-/// Objects that can be stepped over in both directions.
+/// Objects that have a notion of *successor* and *predecessor*
+/// for the purpose of range iterators.
 ///
-/// The `steps_between` function provides a way to efficiently compare
-/// two `Step` objects.
+/// This trait is `unsafe` because implementations of the `unsafe`
+/// trait `TrustedLen` depend on its implementation being correct.
 #[unstable(feature = "step_trait",
-           reason = "likely to be replaced by finer-grained traits",
+           reason = "recently redesigned",
            issue = "42168")]
-pub trait Step: Clone + PartialOrd + Sized {
-    /// Returns the number of steps between two step objects. The count is
-    /// inclusive of `start` and exclusive of `end`.
+pub unsafe trait Step: Clone + PartialOrd + Sized {
+    /// Returns the number of *successor* steps needed to get from `start` to `end`.
     ///
-    /// Returns `None` if it is not possible to calculate `steps_between`
-    /// without overflow.
+    /// Returns `None` if that number would overflow `usize`
+    /// (or is infinite, or if `end` would never be reached).
+    ///
+    /// # Invariants
+    ///
+    /// For any `a`, `b`, and `n`:
+    ///
+    /// * `steps_between(&a, &b) == Some(0)` if and only if `a >= b`
+    /// * `steps_between(&a, &b) == Some(n)` if and only if `a.forward(n) == Some(b)`
+    /// * `steps_between(&a, &b) == Some(n)` if and only if `b.backward(n) == Some(a)`
     fn steps_between(start: &Self, end: &Self) -> Option<usize>;
 
-    /// Replaces this step with `1`, returning itself.
-    fn replace_one(&mut self) -> Self;
-
-    /// Replaces this step with `0`, returning itself.
-    fn replace_zero(&mut self) -> Self;
-
-    /// Adds one to this step, returning the result.
-    fn add_one(&self) -> Self;
+    /// Returns the value that would be obtained by taking the *successor*
+    /// of `self` `count` times.
+    ///
+    /// Returns `None` if this would overflow the range of values supported by `Self`.
+    ///
+    /// # Invariants
+    ///
+    /// For any `a`, `n`, and `m` where `n + m` does not overflow:
+    ///
+    /// * `a.forward(n).and_then(|x| x.forward(m)) == a.forward(n + m)`
+    /// * `a.forward(n)` equals `Step::successor` applied to `a` `n` times
+    fn forward(&self, count: usize) -> Option<Self>;
 
-    /// Subtracts one to this step, returning the result.
-    fn sub_one(&self) -> Self;
+    /// Returns the *successor* of `self`.
+    ///
+    /// If this would overflow the range of values supported by `Self`,
+    /// this method is allowed to panic or overflow. Suggested behavior is
```

Comment:
Note to self: these should probably say "wrap" rather than "overflow"/"underflow".